### ISSUE TITLE
SearchByActorAsync wrong return type

### ISSUE THREAD

**Issue by PoLaKoSz** [tonykaralis/TMdbEasy#5]
In TMDb v3 the [search/person](https://developers.themoviedb.org/3/search/search-people) endpoint response different from the [search/movies](https://developers.themoviedb.org/3/search/search-movies) so the response can not be `Task<MovieList>`.

**Comment by tonykaralis** [comment]
Having looked into this in more detail, this is what it returns

```
public class RootObject
    {
        public int page { get; set; }
        public List<Result> results { get; set; }
        public int total_results { get; set; }
        public int total_pages { get; set; }
    }
```

   ```
 public class Result
    {
        public string profile_path { get; set; }
        public bool adult { get; set; }
        public int id { get; set; }
        public List<KnownFor> known_for { get; set; }
        public string name { get; set; }
        public double popularity { get; set; }
    }
```

```
public class KnownFor
{
    public string poster_path { get; set; }
    public bool adult { get; set; }
    public string overview { get; set; }
    public string release_date { get; set; }
    public string original_title { get; set; }
    public List<object> genre_ids { get; set; }
    public int id { get; set; }
    public string media_type { get; set; }
    public string original_language { get; set; }
    public string title { get; set; }
    public string backdrop_path { get; set; }
    public double popularity { get; set; }
    public int vote_count { get; set; }
    public bool video { get; set; }
    public double vote_average { get; set; }
    public string first_air_date { get; set; }
    public List<string> origin_country { get; set; }
    public string name { get; set; }
    public string original_name { get; set; }
}
```

Passing in an actors name will return all actors that match eg. brad pi may return multiple results. Brad pitt on the other hand will return a single result. Point is if the user types in the correct name then thats ok, but what if they dont??

**Comment by tonykaralis** [comment]
We can either write logic to filter out the actual actor which will be difficult if they have typed it in very incorrectly. or we can build up a list of actors from their search and allow them to pick the correct actor. Then we can perform the search and be gauranteed to get **only 1 result**. We can then extract all the movies from the known for object and add use those as needed

**Comment by PoLaKoSz** [comment]
My idea not fully good because I thought like we would use it in MoodMovies. This library should return the deserialized response and the MoodMovies responsibility to extract only the movies from every actor.

**Comment by tonykaralis** [comment]
Yeah 100% agree. This library will return the results as above

**Comment by PoLaKoSz** [comment]
Hi!

Lucky for Us there is an another problem :smile:

The response `results.(actor object).known_for` list can contains two types of object:
- MovieList type of object and
- TV List type of object

My ideas:
- first I thought it will be good enought to return the `known_for` list as a List\<object\> but this solution will require some nasty code when somebody wants to use the `SearchByActorAsync( ... )` method. Than
- I thought what if We store the results like this:
```` C#
class MixedSearchResults
{
   public List<Movie> Movies { get; }
   public List<TV> TvList{ get; }
}
````

Do You have any other idea?
Maybe the names represented in the code example not the best, Can You give your thoughts about it?
Thanks

**Comment by tonykaralis** [comment]
Hi,

I agree that this is nasty. 

> first I thought it will be good enought to return the known_for list as a List<object> but this solution will require some nasty code when somebody wants to use the SearchByActorAsync( ... ) method. Than

Your second point makes more sense but will that mean that one of those lists will be null when a query is run and return to the user?

Sorry I am a bit detached from this as havent worked with it for a while, did you find this out from the docs on TMDB?

**Comment by PoLaKoSz** [comment]
No, my religion deny using null-s 😄 I against using null-s no matter what. Ctor for initializing everything.

With the second implementation the `MixedSearchResults` class would be the JSON `known_for` array. So the two `List<T>` should be populated (I think this needs a custom Deserializer) and returned from the `SearchByActorAsync( ... )`.

**Comment by tonykaralis** [comment]
Hahahah good one 👍  i think we can add a custom deserializer as long as we dont touch the already existing Json code. We can another method that can deserialize this tricky type. Are you up for the challenge?

**Comment by PoLaKoSz** [comment]
What do You mean by the `already existing Json code`?

Yes, You can assign this to me (but maybe it will take a few days to implement and commit it).

**Comment by tonykaralis** [comment]
Have you had any luck with this this?

**Comment by PoLaKoSz** [comment]
Hi!

I will implement it today.

**Comment by tonykaralis** [comment]
@PoLaKoSz once v1 is merged then I think we can look at fixing this. It could be done using a custom json serialization plugin perhaps but something to look at after v1
